### PR TITLE
B/graphql

### PR DESCRIPTION
**Pull request by worldofjoni**
# Pull Request for initial Implementation

## Backend
Make sure, you have done de following before merging this pull request:
- [x] all todos are resolved
- [x] check whether all graphql attributes are present
- [x] Your code is **documented** using doc comments (_including all Graphql attributes_).
- [x] You added appropriate **logging** in your code.
- [x] You added **unit tests** to all your implemented functions.
- [x] You ran `cargo format` to format your code.
- [x] You ran `cargo clippy` to check for improvements.
- [x] You added the **class descriptions** of the classes you implemented in the **latex document** on overleaf.
      Make sure to mark any changes you made to the design documents specification.

## TODOs
- [x] server logging
- [x] test depth / complexity limit

### CODE REVIEW

File: backend/src/layer/trigger/graphql/mock.rs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+                price_employee: 0,
+                price_guest: 0,
+                price_pupil: 0,
+            },
+            last_served: NaiveDate::default(),
+            next_served: NaiveDate::default(),
+            relative_frequency: 0.0,
+            rating_count: 0,
+            average_rating: 0.0,
+            date: Date::from_ymd_opt(2023, 7, 4).expect("HELP!"),
```
Comment (worldofjoni) [inline backend/src/layer/trigger/graphql/mock.rs:91]: please a more healpful error message

Comment (worldofjoni) [reply backend/src/layer/trigger/graphql/mock.rs:91]: (multiple places)

File: backend/src/layer/trigger/graphql/mutation.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+use super::util::ApiUtil;
+
+/// Class implementing `GraphQLs` root mutations.
+#[derive(Debug)]
+pub struct MutationRoot;
+
+#[Object]
+impl MutationRoot {
+    /// This query adds an image to the specified main dish.
```
Comment (worldofjoni) [inline backend/src/layer/trigger/graphql/mutation.rs:14]: Rename `query` to `mutation`?

File: backend/src/layer/trigger/graphql/types/line.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    /// The name of the line
+    name: String,
+    #[graphql(skip)]
+    /// The id of the canteen to which the line belongs. Currently only used for getting the canteen
+    canteen_id: Uuid,
+}
+
+#[ComplexObject]
+impl Line {
+    /// A function for getting the canteen this line belongs to
```
[the remark this answers is not in the document]
Comment (worldofjoni) [reply backend/src/layer/trigger/graphql/types/line.rs:26]: Multiple places

[the remark this answers is not in the document]
Comment (worldofjoni) [reply backend/src/layer/trigger/graphql/types/line.rs:26]: Also: these resolver functions should not be called "function" as they are just a attribute for a graphql user.

File: backend/src/layer/trigger/graphql/types/line.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    #[instrument(skip(ctx))]
+    #[graphql(complexity = "10 * child_complexity")]
+    async fn canteen(&self, ctx: &Context<'_>) -> Result<Canteen> {
+        // TODO Option when no data available?
```
Comment (worldofjoni) [inline backend/src/layer/trigger/graphql/types/line.rs:30]: this todo comment can be removed

File: backend/src/layer/trigger/graphql/types/meal.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    /// Type of this meal.
+    /// Here the type of meat which is contained in the meal, or whether it is vegetarian or vegan, is specified.
+    meal_type: MealType,
+    /// The ratings given by the users to the meal.
+    ratings: Ratings,
+    /// The prices of the dish each for the four groups of people students, employees, pupils and guests.
+    price: Price,
+    /// Some statistics for the meal.
+    statistics: MealStatistics,
+    #[graphql(skip)]
```
Comment (worldofjoni) [inline backend/src/layer/trigger/graphql/types/meal.rs:29]: the skipped attributes dont need dock comments as they are private and not exposed to the public graphql api.
If they still are wanted to be keept they should be put before the `#` attributes.

Comment (worldofjoni) [reply backend/src/layer/trigger/graphql/types/meal.rs:29]: multiple places

File: backend/src/util.rs [hunk 6]
```diff
@@ -62,9 +65,14 @@ pub enum Allergen {
...
     /// This meal contains molluscs.
     Wt,
+    /// This meal contains animal rennet.
+    La,
+    /// This meal contains gelatin.
+    Gl,
 }
 
 /// This enum lists every possible additive a meal can have.
+#[derive(Copy, Clone, Eq, PartialEq, Enum)]
```
Comment (worldofjoni) [inline backend/src/util.rs:75]: derive debug

**Review by Whatsuup — APPROVED** [review]

**Pull request merged**